src/subcommand.rs

```diff
@@ -54,19 +54,28 @@ impl Subcommand {
         );
 
         // Scan the given and all parent directories for a Cargo.toml containing a workspace
-        // TODO: If set, validate that the found workspace is either the given `--manifest-path`,
-        // or contains the given `--manifest-path` as member.
         let workspace_manifest = utils::find_workspace(&search_path)?;
 
-        let (manifest_path, manifest) = if let Some((workspace_manifest_path, workspace)) =
-            &workspace_manifest
-        {
-            // If a workspace was found, find packages relative to it
-            utils::find_package_manifest_in_workspace(workspace_manifest_path, workspace, package)?
-        } else {
-            // Otherwise scan up the directories
-            utils::find_package_manifest(&search_path, package)?
-        };
+        let (manifest_path, manifest) =
+            if let (Some(package), Some((workspace_manifest_path, workspace))) =
+                (package, &workspace_manifest)
+            {
+                // If a workspace was found, and the user chose a package with `-p`, find packages relative to it
+                // TODO: What if we call `cargo apk run` in the workspace root, and detect a workspace? It should
+                // then use the `[package]` defined in the workspace (will be found below, though, but currently)
+                // fails with UnexpectedWorkspace)
```

Comment:
For this to work I should simply remove

```rust
    // This function shouldn't be called when a workspace exists.
    if manifest.workspace.is_some() {
        return Err(Error::UnexpectedWorkspace(manifest_path));
    }
```

but somehow that doesn't feel all too robust :/

Reply:
Ended up implementing this by explicitly comparing a search path (if the user doesn't specify a package name with `-p`) based on `$PWD` or `--manifest-path` against the found workspace (also found based on to `--manifest-path` or `$PWD`) members. Should be more robust/correct, but I hope it doesn't miss any edge-cases.